torch/quantization/_equalize.py

```diff
@@ -0,0 +1,115 @@
+import torch
+import torch.nn as nn
+import copy
+
+def channel_range(input, axis=0):
+    ''' finds the range of weights associated with a specific channel
+
+    '''
+    size_of_tensor_dim = len(input.size())
+    mins = input
+    maxs = input
+    for i in range(size_of_tensor_dim - 1, axis, -1):
+        mins = torch.min(mins, i)[0]
+        maxs = torch.max(maxs, i)[0]
+    for i in range(axis):
+        mins = torch.min(mins, 0)[0]
+        maxs = torch.max(maxs, 0)[0]
+    return maxs.sub(mins)
+
+def scaling_channels(tensor1, tensor2, output_axis=0, input_axis=1):
+    ''' Given two adjacent tensors', the weights are scaled such that
+    the ranges of the first tensors' output channel are equal to the
+    ranges of the second tensors' input channel
+
+    Note: assumes the input/output channels are the first two axises of the given tensors
+    '''
+    if tensor1.size()[output_axis] != tensor2.size()[input_axis]:
+        raise TypeError("Incompatible tensors")
+
+    output_channel_tensor1 = channel_range(tensor1, output_axis)
+    input_channel_tensor2 = channel_range(tensor2, input_axis)
+
+    # producing scaling factors to applied
+    # zero_input = torch.zeros(output_channel_tensor1.size())
+    # scaling_factors = torch.sqrt(torch.addcdiv(zero_input, output_channel_tensor1, input_channel_tensor2))
+    input_channel_tensor2 += 1e-9
+    scaling_factors = torch.sqrt(output_channel_tensor1 / input_channel_tensor2)
+    r_scaling_factors = torch.reciprocal(scaling_factors)
+
+    # formatting the scaling (1D) tensors to be applied on the given argument tensors
+    scaling_factors = torch.unsqueeze(scaling_factors, output_axis)
+    r_scaling_factors = torch.unsqueeze(r_scaling_factors, input_axis)
+
+    for i in range(len(tensor1.size()) - 2):
+        r_scaling_factors = torch.unsqueeze(r_scaling_factors, -1)
+    for i in range(len(tensor2.size()) - 2):
+        scaling_factors = torch.unsqueeze(scaling_factors, -1)
+
+    # mod_tensor1 = torch.mul(tensor1, r_scaling_factors)
+    # mod_tensor2 = torch.mul(tensor2, scaling_factors)
+    mod_tensor1 = tensor1 * r_scaling_factors
+    mod_tensor2 = tensor2 * scaling_factors
+
+    return mod_tensor1, mod_tensor2
+
+
+def cross_layer_equalization(module1, module2):
```

Comment:
equalization is only applied on weights, I think it's better to have this function only accept weights, you can do verification outside of this function.

Also looks like for both conv and linear the channel axis for first module is always 0 and always 1 for second module, so we don't really need to check that. Although you can do some validation checking the dimension of these axis match.

Reply:
nit: we want to use verbs for functions that does something, so: cross_layer_equalization --> `cross_layer_equalize` or just `_equalize_impl`